runtime/server/generate_chart.go

```diff
@@ -0,0 +1,174 @@
+		if err != nil {
+			return nil, err
+		}
+		defer release()
+		olap, ok := handle.AsOLAP(req.InstanceId)
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "connector is not an OLAP connector")
+		}
+
+		// Get table info
+		tbl, err := olap.InformationSchema().Lookup(ctx, req.Table)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, "table not found")
+		}
+
+		schema = tbl.Schema
+		table = req.Table
+	} else {
+		// TODO: support metrics view
+		return nil, errors.New("at lest one of chart or table must be specified")
+	}
+
+	resp, err := s.generateChartWithAI(ctx, req.InstanceId, table, req.Prompt, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the vega lite spec json to string
+	spec, err := json.Marshal(resp.VegaLiteSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtimev1.GenerateChartSpecResponse{
+		VegaLiteSpec: string(spec),
+		Sql:          resp.SQL,
+	}, nil
+}
+
+type chartAIResponse struct {
+	VegaLiteSpec map[string]interface{} `json:"vega_lite_spec"`
+	SQL          string                 `json:"sql"`
+}
+
+// generateChartWithAI attempts to generate a vega lite chart spec and a SQL used to display the chart.
+func (s *Server) generateChartWithAI(ctx context.Context, instanceID, tblName, userPrompt string, schema *runtimev1.StructType) (*chartAIResponse, error) {
+	// Build messages
+	msgs := []*drivers.CompletionMessage{
+		{Role: "system", Data: chartYAMLSystemPrompt()},
+		{Role: "user", Data: chartYAMLUserPrompt(tblName, userPrompt, schema)},
+	}
+
+	// Connect to the AI service configured for the instance
+	ai, release, err := s.runtime.AI(ctx, instanceID)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
+
+	// Apply timeout
+	ctx, cancel := context.WithTimeout(ctx, aiGenerateTimeout)
+	defer cancel()
+
+	// Call AI service to infer a metrics view YAML
+	res, err := ai.Complete(ctx, msgs)
+	if err != nil {
+		return nil, err
+	}
+
+	// The AI may produce Markdown output. Remove the code tags around the YAML.
+	res.Data = strings.TrimPrefix(res.Data, "```yaml")
+	res.Data = strings.TrimPrefix(res.Data, "```")
+	res.Data = strings.TrimSuffix(res.Data, "```")
+
+	res.Data = strings.TrimPrefix(res.Data, "```json")
+	res.Data = strings.TrimPrefix(res.Data, "```")
+	res.Data = strings.TrimSuffix(res.Data, "```")
+
+	res.Data = strings.TrimPrefix(res.Data, "```sql")
+	res.Data = strings.TrimPrefix(res.Data, "```")
+	res.Data = strings.TrimSuffix(res.Data, "```")
+
+	resp := &chartAIResponse{}
+	err = json.Unmarshal([]byte(res.Data), resp)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// chartYAMLSystemPrompt returns the static system prompt for the chart generation AI.
+func chartYAMLSystemPrompt() string {
+	return `
+You are an agent whose only task is to suggest relevant chart based on a table schema.
+You should suggest the vega config for the chart based on the data.
+Replace the data field in vega lite json with,
+{ "name": "table" }
```

Comment:
With metrics view generation, it seemed that less fixed rule were better. So maybe consider saying "don't add a 'data' field". And then just add this programmatically afterwards.

Reply:
One thing I am seeing is chatgpt does a good job of giving proper indentation. Parsing, updating and reformatting will lose this. So probably it might be good to keep it as is.

Reply:
Okay, you can keep if you want. But note that Go's JSON encoder is able to output indented JSON.

Reply:
While true, I dont think any library can do partial indentation like,
```
 {
    "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
    "data": { "name": "table" },
    "mark": "arc",
    "encoding": {
      "theta": {"field": "total_records", "type": "quantitative"},
      "color": {"field": "publisher", "type": "nominal"}
    },
    "width": 300,
    "height": 300
  }
```

Reply:
That's probably true :) Makes sense